bigomics/omicsplayground
# Testing of Omics Playground

## Omics Playground: Automatic testing routines aimed at continuous integration

### How to run automated testing
Before PRs, you should always runTests to identify issues with boards.

1. Similar to an R package, go to root ("omicsplayground/") and run `Rscript -e "shiny::runTests()"`
2. You will get instructions on the console on how to proceed if there is an error.

In summary, snapshorts will be saved in `tests/testthat/_snaps` and these will be used as a reference. You should always check these snaps before submitting your PR.

### Automated checks of pgx file error
We can now provide a target folder with multiple pgx files (target_folder), and have all pgx files inside the folder (recursive) checked for error.

1. To check pgx files for error, provide the target folder in the script `Rscript dev//board_check_across_pgx.R -d target_folder`
2. Alternatively, you can run makefile target check_check as in `make check_pgx d=data/mini-example`

As a result, a file `pgx_check_results.csv` will be created in the target folder. As seen bellow, the board PCSF has an error when running mini example data.

<img width="947" alt="image" src="https://github.com/bigomics/omicsplayground/assets/28757711/e51cb887-cd38-48ac-a98d-b0ffad1e3127">
<br></br>
This is confirmed by looking at the board:
<br></br>
<img width="687" alt="image" src="https://github.com/bigomics/omicsplayground/assets/28757711/ef52107a-2bdf-4999-836c-298f2d316b4a">

### How to run a single board

We have two targets in make file. In both cases, we need to specify the board. 

1. `make board` will start a board without dataset. You should specify a board, for example `make board board=dataview`
3. `make board_example` will start a board with the example dataset. You should specify a board, for example `make board_example board=dataview`

### Authentication in single board

We can now run authentication by providing argument `auth`, as in `make board2 board=enrichment auth=none`, default is NULL, so it will not run authentication board. Same values for authentication available in `OPTIONS` file are valid.


### Changes necessary to implement testing

1. golem had to do, as shinyapp path was not compatible with shinytest2
2. single board code was completely refactored
4. board.launch runs from a single script (ui, server) in dev/board.launch, instead of multiple as happened before
5. single board launch is now completely compliant to the main App, with same global.R, ui and server structure.
6. several changes in boards had to be made to make testing robust
7. some functional functions had to be converted to pathway to make single board launch happens
8. mini example data had to be added. Thanks @phisanti! Mini example data will have two functions in this PR:
> - be used in all board launches/tests with option to use example data, so we launch/test them quicker;
> - be used as comparison dataset for board.comparison

### Next steps
1. Three boards our of 18 had problems with reactivity/logic and this was caught in testing: `c("featuremap","intersection", "wordcloud")`
2. PCSF has an error in mini example data.
3. Implement missing complex boards `c("upload", "loading")`
4. Create additional check routines
> - upload and compute example dataset
>- download plot, table, etc...
5. Add a github action for automated testing
